src/fluree/db/json_ld/shacl.cljc

```diff
@@ -303,45 +323,115 @@
            (recur r path-flakes*))
          path-flakes)))))
 
+(declare build-property-shape)
+(defn validate-qualified-shape-constraints
+  "Takes a property shape with a qualifiedValueShape constraint, builds the shape,
+  validates it, and returns the shape with all conforming sids."
+  [db {:keys [path qualified-value-shape qualified-min-count qualified-max-count qualified-value-shapes-disjoint] :as p-shape}
+   p-flakes]
+  (go-try
+    (let [qualified-shape (<? (build-property-shape db const/$sh:qualifiedValueShape qualified-value-shape))]
+      (loop [[f & r] p-flakes
+             conforming    #{}]
+        (if f
+          (let [sid           (flake/o f)
+                s-flakes      (<? (query-range/index-range db :spot = [sid]))
+                pid->p-flakes (group-by flake/p s-flakes)
+                path-flakes   (<? (resolve-path-flakes db sid (:path qualified-shape) pid->p-flakes))
+                [valid? :as res] (<? (validate-property-constraints qualified-shape path-flakes db))]
+            (recur r (if valid?
+                       (conj conforming sid)
+                       conforming)))
+          (assoc p-shape :conforming conforming))))))
+
+(defn validate-qualified-cardinality-constraints
+  [{:keys [path conforming qualified-min-count qualified-max-count]}]
+  (let [conforming-count (count conforming)]
+    (cond (and qualified-min-count (< conforming-count qualified-min-count))
+          [false (str "path " path " conformed to sh:qualifiedValueShape fewer than sh:qualifiedMinCount times")]
+
+          (and qualified-max-count (> conforming-count qualified-max-count))
+          [false (str "path " path " conformed to sh:qualifiedValueShape more than sh:qualifiedMaxCount times")]
+
+          :else
+          [true nil])))
```

Comment:
The `true` outcomes usually need error messages too as those are used when the whole thing is wrapped in a `sh:not`. But does this one not need that for some reason?

Reply:
Ah, I wondered why we had error messages for valid results, that makes sense. I'll add them.